src/workerd/api/global-scope.c++

```diff
@@ -398,10 +398,14 @@ kj::Promise<WorkerInterface::AlarmResult> ServiceWorkerGlobalScope::runAlarm(
       auto& persistent = KJ_ASSERT_NONNULL(actor.getPersistent());
       persistent.cancelDeferredAlarmDeletion();
 
-      if (auto desc = e.getDescription();
-          !jsg::isTunneledException(desc) && !jsg::isDoNotLogException(desc)) {
-        LOG_EXCEPTION("alarmRun"_kj, e);
-      }
+      // This will include the error in inspector/tracers and log to syslog if internal.
+      context.logUncaughtExceptionAsync(UncaughtExceptionSource::ALARM_HANDLER, kj::mv(e));
+
+      // TODO(someday) We don't invoke `RequestObserver::reportFailure()` here because we do not
+      // have access to a `RequestObserver` in this layer. If we can move the translation from
```

Comment:
Stupid question, why can't we access `RequestObserver` here? Wouldn't we still have an `IncomingRequest` here since an alarm is finishing up, i.e. `context.getMetrics().reportFailure()`?

Reply:
That's a good question! So we do have an `IncomingRequest` on the stack in `WorkerEntrypoint`. But we don't pass it in to any of the functions in `ServiceWorkerGlobalScope`. The part I care about here is that `IncomingRequest` is part of the `io` layer and `ServiceWorkerGlobalScope` is part of the `api` layer. This is loosely motivated by something called [Separation of Concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). (See also https://en.wikipedia.org/wiki/Law_of_Demeter.) Basically, layers are informal boundaries between components to make them easier to maintain. We don't want to care too much about external scheduling and networking while working in JS and we don't want to care about v8 in our scheduling.

All of that said, this question made me realize that we do use the `IoContext` in the `api` layer and it does have access to the `RequestObserver`!

Reply:
Thanks, appreciate the clarification!